lib/ppx/ppx_deriving_json.ml

```diff
@@ -0,0 +1,597 @@
+        write_body_of_tuple_type l ~arg ~poly ~tag:0
+    in
+    Ast_helper.Exp.case lhs rhs
+  | Rinherit ({ptyp_desc = (Ptyp_constr (lid, _) as c)} as y) ->
+    Ast_helper.Exp.case (Ast_helper.Pat.type_ lid)
+      (write_body_of_type y ~arg ~poly)
+
+and write_body_of_type y ~arg ~poly =
+  match y with
+  | [%type: unit] ->
+    [%expr Deriving_Json.Json_unit.write buf [%e arg]]
+  | [%type: int] ->
+    [%expr Deriving_Json.Json_int.write buf [%e arg]]
+  | [%type: int32] | [%type: Int32.t] ->
+    [%expr Deriving_Json.Json_int32.write buf [%e arg]]
+  | [%type: int64] | [%type: Int64.t] ->
+    [%expr Deriving_Json.Json_int64.write buf [%e arg]]
+  | [%type: nativeint] | [%type: Nativeint.t] ->
+    [%expr Deriving_Json.Json_nativeint.write buf [%e arg]]
+  | [%type: float] ->
+    [%expr Deriving_Json.Json_float.write buf [%e arg]]
+  | [%type: bool] ->
+    [%expr Deriving_Json.Json_bool.write buf [%e arg]]
+  | [%type: char] ->
+    [%expr Deriving_Json.Json_char.write buf [%e arg]]
+  | [%type: string] ->
+    [%expr Deriving_Json.Json_string.write buf [%e arg]]
+  | [%type: bytes] ->
+    [%expr Deriving_Json.Json_bytes.write buf [%e arg]]
+  | [%type: [%t? y] list] ->
+    let e = write_of_type y ~poly in
+    [%expr Deriving_Json.write_list [%e e] buf [%e arg]]
+  | [%type: [%t? y] ref] ->
+    let e = write_of_type y ~poly in
+    [%expr Deriving_Json.write_ref [%e e] buf [%e arg]]
+  | [%type: [%t? y] option] ->
+    let e = write_of_type y ~poly in
+    [%expr Deriving_Json.write_option [%e e] buf [%e arg]]
+  | [%type: [%t? y] array] ->
+    let e = write_of_type y ~poly in
+    [%expr Deriving_Json.write_array [%e e] buf [%e arg]]
+  | { Parsetree.ptyp_desc = Ptyp_var v } when poly ->
+    [%expr [%e Ast_convenience.evar ("poly_" ^ v)] buf [%e arg]]
+  | { Parsetree.ptyp_desc = Ptyp_tuple l } ->
+    write_body_of_tuple_type l ~arg ~poly ~tag:0
+  | { Parsetree.ptyp_desc = Ptyp_variant (l, _, _); ptyp_loc = loc } ->
+    List.map (write_poly_case ~arg ~poly) l @ [unreachable_case ()] |>
+    Ast_helper.Exp.match_ arg
+  | { Parsetree.ptyp_desc = Ptyp_constr (lid, l) } ->
+    let e = suffix_lid lid ~suffix:"to_json"
+    and l = List.map (write_of_type ~poly) l in
+    [%expr [%e Ast_convenience.app e l] buf [%e arg]]
+  | { Parsetree.ptyp_loc } ->
+    Location.raise_errorf ~loc:ptyp_loc
+      "%s_write cannot be derived for %s"
+      deriver (Ppx_deriving.string_of_core_type y)
+
+and write_of_type y ~poly =
+  let v = "a" in
+  let arg = Ast_convenience.evar v
+  and pattern = Ast_convenience.pvar v in
+  wrap_write (write_body_of_type y ~arg ~poly) ~pattern
+
+and write_of_record d l =
+  let pattern =
+    let l =
+      let f {Parsetree.pld_name} =
+        label_of_constructor pld_name,
+        Ast_helper.Pat.var pld_name
+      in
+      List.map f l
+    in
+    (* CHECKME: what is the closed_flag for? *)
+    Ast_helper.Pat.record l Asttypes.Closed
```

Comment:
This is to differentiate between `{ x }` and `{ x; _ }`. The former has an exhaustiveness warning, turned off by default.